### ISSUE TITLE
Layered deployments are not working

### ISSUE THREAD

**Issue by lerppana** [Azure/azure-iot-cli-extension#147]
This is problem that could be fixed from multiple sides, but I´m gonna post this here:

We have a product with which we would like to use layered deployments. This is the template.json that I use to generate the deployment from:

(generalized some company names)
```json
{
    "$schema-template": "2.0.0",
    "modulesContent": {
        "$edgeAgent": 
        {
            "properties.desired": 
            {
                "modules": 
                {
                   "machine-learning-module": 
                   {
                        "type": "docker",
                        "status": "running",
                        "restartPolicy": "always",
                        "version": "1.0",
                        "settings": 
                        {
                            "image": "${MODULEDIR<modules/MachineLearningModule>}",
                            "createOptions":{
                                "HostConfig": 
                                {
                                    "Binds": [ "${IOTEDGE_SHARED_ROOT}/iotedge:${IOTEDGE_SHARED_ROOT_CONTAINER}/" ]
                                }
                            }
                        }
                    }
                },
                "runtime":{
                    "settings": {
                        "registryCredentials": {
                            "devcontainer": {
                                "username": "${ACR_USER}",
                                "password": "${ACR_PASS}",
                                "address": "devcontainer.azurecr.io"
                              }
                        }
                    }
                }
            }
        },
        "$edgeHub": 
        {
            "properties.desired.routes.BroadcastToMachineLearning": "FROM /messages/modules/* WHERE Output = 'broadcast' AND $connectionModuleId != 'machine-learning-module' INTO BrokeredEndpoint(\\\"/modules/machine-learning-module/inputs/broadcast\\\")",
            "properties.desired.routes.DataInputToMachineLearning": "FROM /messages/modules/* WHERE Output = 'data' AND Module = 'machine-learning-module' INTO BrokeredEndpoint(\\\"/modules/machine-learning-module/inputs/dataInput\\\")"
        }
    }
}
``` 

This template can be generated to a deployment.json with vscode. Unfortunately, when using CLI to deploy this into IotHub it will not be layered deployment but a normal deployment. I got this to work by changing the template.json to be of a format:
```json
"$edgeAgent": {
         "properties.desired.runtime.settings.registryCredentials.devcontainer": {
          ...
          },
          "properties.desired.modules.machine-learning-module": {
          ....
          }
}
```

But that is not a format that is supported by generators. Could this be fixed by CLI, API or should I move this issue to vscode tools?

**Comment by digimaun** [comment]
Thanks for submitting the issue @lerppana .

I am assuming you are using Portal to validate whether a deployment is layered or not. The only mandatory field in a layered deployment is `$edgeAgent`. The portal examines the `$edgeAgent` object. If `properties.desired: {}` is defined the UI treats the deployment as a full edge deployment. If `properties.desired.x` is defined the UI treats the deployment as layered.

In your original example `properties.desired: {}` is defined and thus appears to the UI as a full deployment.

Bottom line, `$edgeAgent` and `$edgeHub` must both be either layered or full deployment format. They cannot be mixed-and-matched. You discovered this when providing the reformatted template.json version.

One enhancement we could add to the CLI would be a guard to prevent you from submitting the original format as layered.

I'd recommend creating an issue for vscode tools to update the template generation logic of a layered deployment.

**Comment by lerppana** [comment]
Thanks for the answer! Indeed it would be beneficial to have some sort of reason given for user that explains what happened after deployment. Even though I used switch `--layered` deployment went through as normal deployment without any warnings because I did not use correct format! Created some confusion. I will close this issue and pursue this issue on tools.
